improver/psychrometric_calculations/hail_size.py

```diff
@@ -0,0 +1,295 @@
+# AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+# IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
+# ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE
+# LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
+# CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
+# SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
+# INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
+# CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+# ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+# POSSIBILITY OF SUCH DAMAGE.
+
+"""modules to calculate hail_size"""
+
+import math
+from operator import itemgetter
+from typing import List
+
+import iris
+import numpy as np
+from iris.cube import Cube
+
+from improver import BasePlugin
+from improver.metadata.utilities import (
+    create_new_diagnostic_cube,
+    generate_mandatory_attributes,
+)
+from improver.psychrometric_calculations.psychrometric_calculations import (
+    adjust_for_latent_heat,
+    dry_adiabatic_temperature,
+    saturated_humidity,
+)
+from improver.utilities.cube_manipulation import MergeCubes
+
+
+class HailSize(BasePlugin):
+    def __init__(self, model_id_attr: str = None):
+        self.model_id_attr = model_id_attr
+        self.ccl_temperature = None
+        self.ccl_pressure = None
+
+    def nomogram_values(self):
+
+        """set-up dictionary of the table found in https://doi.org/10.1002/met.236 of hail sizes"""
+
+        """formatting turned off for this graph as the formatting makes it unreadable. Might need to 
+        reconsider how I'm formatting this"""
+
+        # fmt: off
+        lookup_nomogram={
+            5:  [0 ,0 ,0 ,0 ,0 ,0 ,0 ,2 ,2 ,2 ,2 ,2 ,5 ,5 ,5 ,5 ,10,10,10,10 ,10 ,10 ,10 ,15 ,15 ,15 ],
+            10: [0 ,0 ,0 ,0 ,0 ,2 ,2 ,2 ,5 ,5 ,5 ,10,10,15,15,15,15,15,15,20 ,20 ,20 ,20 ,25 ,25 ,25 ],
+            15: [0 ,0 ,2 ,2 ,2 ,2 ,5 ,10,10,15,15,20,20,20,25,25,25,25,30,30 ,30 ,35 ,35 ,35 ,35 ,35 ],
+            20: [2 ,2 ,2 ,2 ,2 ,5 ,10,15,20,20,20,25,25,30,30,35,35,40,40,40 ,40 ,45 ,45 ,45 ,45 ,50 ],
+            25: [2 ,5 ,5 ,5 ,10,15,20,20,25,30,30,35,40,40,40,40,45,45,45,50 ,50 ,50 ,50 ,50 ,55 ,55 ],
+            30: [5 ,5 ,5 ,10,15,20,20,25,30,35,40,40,40,45,45,50,50,55,55,60 ,60 ,60 ,60 ,65 ,65 ,65 ],
+            35: [5 ,5 ,10,15,20,20,25,30,35,40,40,45,45,50,55,55,60,60,65,70 ,70 ,75 ,75 ,80 ,80 ,80 ],
+            40: [5 ,5 ,10,15,20,20,25,30,35,40,40,45,50,55,60,60,65,70,75,80 ,80 ,85 ,90 ,100,100,100],
+            45: [5 ,10,15,20,20,25,30,35,40,45,45,50,55,60,60,65,70,80,85,90 ,95 ,100,105,110,110,110],
+            50: [5 ,10,15,20,20,25,30,35,40,45,50,50,55,60,65,75,80,85,90,100,105,110,115,120,120,120]  
+        }
+        # fmt: on
+
+        return lookup_nomogram
+
+    def temperature_of_atmosphere_at_CCL_pressure(
```

Comment:
As with the method below, this is ferociously slow. Here's an alternative that is a bit more sprightly. It's worth you considering these, they may not be as efficient as they could be, but hopefully they offer something to learn from.

```
def find_temperature_at_pressure(pressure, temperature):
    t_at_p = iris.cube.CubeList()
    
    pressure_coord = build_pressure_ref(temperature)
    
    for pslice, tslice in zip(pressure.slices_over("realization"), temperature.slices_over("realization")): # Working realization by realization to limit memory use.
        template = temperature_slice(tslice)  # A temperature template containins the correct realization scalar value.
        pdiff = abs(pressure_coord - pslice.data)  # Calculating the absolute pressure difference between the coordinate values and the CCL value.
        indices = np.nanargmin(pdiff, axis=0)  # Finding the smallest difference (assumes pressure always decreases as we go up).
        m,n = indices.shape  
        I,J = np.ogrid[:m,:n]  # Generate y,x coords from the shape of the returned 2D array. 
        template.data = tslice.data[indices, I, J]  # Index the 3D temperature data using the pressure, y, and x index values.
        t_at_p.append(template)  # Store the result for each realization.
        
    return t_at_p.merge_cube()  # Merge the slices to reconstruct the realization coordinate.

def build_pressure_ref(temperature):
    pressure_coord = temperature.coord("pressure").points
    template = next(temperature.slices_over(["realization", "pressure"]))
    ref = []
    for point in pressure_coord:
        ref.append(np.broadcast_to(point, template.shape))  # Broadcast the pressure coordinate values to a 3D array the same shape as the temperature grid.
    return np.array(ref)

def temperature_slice(tslice):
    template = next(tslice.slices_over("pressure")).copy()
    template.remove_coord("pressure")
    return template

```